File: nuget/Antybrowser.SDK/AntybrowserClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Antybrowser.SDK.Models;
using Antybrowser.SDK.Exceptions;

namespace Antybrowser.SDK
{
    /// <summary>
    /// Antybrowser .NET SDK Client for interacting with the Antybrowser Local API.
    /// Provide seamless integration for browser automation, multi-accounting, and anti-detect profile management.
    /// </summary>
    public class AntybrowserClient : IDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly string _apiKey;
        private readonly string _baseUrl;

        /// <summary>
        /// Initializes a new instance of the <see cref="AntybrowserClient"/> class.
        /// </summary>
        /// <param name="apiKey">The API Key found in Antybrowser settings.</param>
        /// <param name="baseUrl">The base URL of the Antybrowser Local API (defaults to http://127.0.0.1:5173).</param>
        public AntybrowserClient(string apiKey, string baseUrl = "http://127.0.0.1:5173")
        {
            _apiKey = apiKey ?? throw new ArgumentNullException(nameof(apiKey));
            _baseUrl = baseUrl.TrimEnd('/');
            _httpClient = new HttpClient();
            _httpClient.DefaultRequestHeaders.Add("x-api-key", _apiKey);
        }

        #region System
        /// <summary>
        /// Checks the status of the Antybrowser Local API.
        /// </summary>
        public async Task<ApiResponse> GetStatusAsync()
        {
            return await GetAsync<ApiResponse>("/api/status");
        }

        /// <summary>
        /// Gets the current synchronization status for all profiles.
        /// </summary>
        public async Task<SyncStatus> GetSyncStatusAsync()
        {
            return await GetAsync<SyncStatus>("/api/sync/status");
        }

        /// <summary>
        /// Triggers a synchronization refresh, optionally for a specific profile.
        /// </summary>
        /// <param name="profileId">Optional profile ID to sync.</param>
        public async Task<ApiResponse> RefreshSyncAsync(string? profileId = null)
        {
            var body = profileId != null ? new { profileId } : null;
            return await PostAsync<ApiResponse>("/api/sync/refresh", body);
        }

        /// <summary>
        /// Retrieves the global Antybrowser settings.
        /// </summary>
        public async Task<Settings> GetSettingsAsync()
        {
            return await GetAsync<Settings>("/api/settings");
        }
        #endregion

        #region Profiles
        /// <summary>
        /// Lists all available browser profiles.
        /// </summary>
        public async Task<List<Profile>> GetProfilesAsync()
        {
            return await GetAsync<List<Profile>>("/api/profiles");
        }

        /// <summary>
        /// Launches a specific browser profile.
        /// </summary>
        /// <param name="id">The unique profile ID.</param>
        public async Task<ApiResponse<object>> StartProfileAsync(int id)
        {
            return await PostAsync<ApiResponse<object>>($"/api/profiles/{id}/start", null);
        }

        /// <summary>
        /// Stops a currently running browser profile.
        /// </summary>
        /// <param name="id">The unique profile ID.</param>
        public async Task<ApiResponse> StopProfileAsync(int id)
        {
            return await PostAsync<ApiResponse>($"/api/profiles/{id}/stop", null);
        }

        /// <summary>
        /// Creates a new anti-detect browser profile.
        /// </summary>
        /// <param name="request">The profile configuration parameters.</param>
        public async Task<Profile> CreateProfileAsync(CreateProfileRequest request)
        {
            return await PostAsync<Profile>("/api/profiles", request);
        }

        /// <summary>
        /// Duplicates an existing browser profile.
        /// </summary>
        /// <param name="id">Source profile ID.</param>
        /// <param name="name">New profile name (optional).</param>
        /// <param name="directoryName">New directory name (optional).</param>
        public async Task<Profile> DuplicateProfileAsync(int id, string? name = null, string? directoryName = null)
        {
            var body = new { name, directoryName };
            return await PostAsync<Profile>($"/api/profiles/{id}/duplicate", body);
        }

        /// <summary>
        /// Updates the configuration of an existing profile.
        /// </summary>
        /// <param name="id">Profile ID to update.</param>
        /// <param name="updateData">Data to update.</param>
        public async Task<Profile> UpdateProfileAsync(int id, object updateData)
        {
            return await PutAsync<Profile>($"/api/profiles/{id}", updateData);
        }

        /// <summary>
        /// Deletes a browser profile permanently.
        /// </summary>
        /// <param name="id">Profile ID to delete.</param>
        public async Task<ApiResponse> DeleteProfileAsync(int id)
        {
            return await DeleteAsync<ApiResponse>($"/api/profiles/{id}");
        }
        #endregion

        #region Automations
        /// <summary>
        /// Lists all available automation scripts.
        /// </summary>
        public async Task<List<Automation>> GetAutomationsAsync()
        {
            return await GetAsync<List<Automation>>("/api/automations");
        }

        /// <summary>
        /// Executes an automation workflow on a specific profile.
        /// </summary>
        /// <param name="automationId">The automation script ID.</param>
        /// <param name="request">Execution parameters and variables.</param>
        public async Task<RunAutomationResult> RunAutomationAsync(int automationId, RunAutomationRequest request)
        {
            return await PostAsync<RunAutomationResult>($"/api/automations/{automationId}/run", request);
        }
        #endregion

        #region Groups
        /// <summary>
        /// Lists all profile groups.
        /// </summary>
        public async Task<List<Group>> GetGroupsAsync()
        {
            return await GetAsync<List<Group>>("/api/groups");
        }

        /// <summary>
        /// Creates a new group for organizing profiles.
        /// </summary>
        public async Task<Group> CreateGroupAsync(CreateGroupRequest request)
        {
            return await PostAsync<Group>("/api/groups", request);
        }

        /// <summary>
        /// Updates an existing group's information.
        /// </summary>
        public async Task<Group> UpdateGroupAsync(int id, object updateData)
        {
            return await PutAsync<Group>($"/api/groups/{id}", updateData);
        }

        /// <summary>
        /// Deletes a group permanently.
        /// </summary>
        public async Task<ApiResponse> DeleteGroupAsync(int id)
        {
            return await DeleteAsync<ApiResponse>($"/api/groups/{id}");
        }
        #endregion

        #region Proxies
        /// <summary>
        /// Lists all managed proxies.
        /// </summary>
        public async Task<List<Proxy>> GetProxiesAsync()
        {
            return await GetAsync<List<Proxy>>("/api/proxies");
        }

        /// <summary>
        /// Adds a new proxy to the Antybrowser proxy manager.
        /// </summary>
        public async Task<Proxy> CreateProxyAsync(CreateProxyRequest request)
        {
            return await PostAsync<Proxy>("/api/proxies", request);
        }

        /// <summary>
        /// Validates a proxy connection and detects its location and IP.
        /// </summary>
        public async Task<ProxyCheckResult> CheckProxyAsync(object proxyData)
        {
            return await PostAsync<ProxyCheckResult>("/api/proxies/check", proxyData);
        }

        /// <summary>
        /// Performs a bulk validation of multiple proxies.
        /// </summary>
        public async Task<ApiResponse<List<ProxyCheckResult>>> CheckProxiesBulkAsync(List<object> proxies)
        {
            return await PostAsync<ApiResponse<List<ProxyCheckResult>>>("/api/proxies/check-bulk", new { proxies });
        }

        /// <summary>
        /// Deletes a proxy configuration.
        /// </summary>
        public async Task DeleteProxyAsync(int id)
        {
            var response = await _httpClient.DeleteAsync($"{_baseUrl}/api/proxies/{id}");
            if (!response.IsSuccessStatusCode)
            {
                var content = await response.Content.ReadAsStringAsync();
                throw new AntybrowserException("Failed to delete proxy", response.StatusCode, content);
            }
        }
        #endregion

        #region Extensions
        /// <summary>
        /// Lists all uploaded browser extensions.
        /// </summary>
        public async Task<List<Extension>> GetExtensionsAsync()
        {
            return await GetAsync<List<Extension>>("/api/extensions");
        }

        /// <summary>
        /// Deletes a browser extension.
        /// </summary>
        public async Task DeleteExtensionAsync(int id)
        {
            var response = await _httpClient.DeleteAsync($"{_baseUrl}/api/extensions/{id}");
            if (!response.IsSuccessStatusCode)
            {
                var content = await response.Content.ReadAsStringAsync();
                throw new AntybrowserException("Failed to delete extension", response.StatusCode, content);
            }
        }

        /// <summary>
        /// Gets extensions assigned to a specific profile.
        /// </summary>
        /// <param name="profileId">Target profile ID.</param>
        /// <param name="details">Whether to include full extension metadata.</param>
        public async Task<List<Extension>> GetProfileExtensionsAsync(int profileId, bool details = false)
        {
            return await GetAsync<List<Extension>>($"/api/profiles/{profileId}/extensions?details={details.ToString().ToLower()}");
        }

        /// <summary>
        /// Assigns a set of extensions to a profile.
        /// </summary>
        /// <param name="profileId">Target profile ID.</param>
        /// <param name="extensionIds">List of extension IDs to assign.</param>
        public async Task<ApiResponse> SetProfileExtensionsAsync(int profileId, List<int> extensionIds)
        {
            return await PostAsync<ApiResponse>($"/api/profiles/{profileId}/extensions", new { extensionIds });
        }
        #endregion

        #region HTTP Helpers
        private async Task<T> GetAsync<T>(string path)
        {
            var response = await _httpClient.GetAsync($"{_baseUrl}{path}");
            return await HandleResponseAsync<T>(response);
        }

        private async Task<T> PostAsync<T>(string path, object? body)
        {
            var json = body != null ? JsonSerializer.Serialize(body) : "{}";
            var content = new StringContent(json, Encoding.UTF8, "application/json");
            var response = await _httpClient.PostAsync($"{_baseUrl}{path}", content);
            return await HandleResponseAsync<T>(response);
        }

        private async Task<T> PutAsync<T>(string path, object body)
        {
            var json = JsonSerializer.Serialize(body);
            var content = new StringContent(json, Encoding.UTF8, "application/json");
            var response = await _httpClient.PutAsync($"{_baseUrl}{path}", content);
            return await HandleResponseAsync<T>(response);
        }

        private async Task<T> DeleteAsync<T>(string path)
        {
            var response = await _httpClient.DeleteAsync($"{_baseUrl}{path}");
            return await HandleResponseAsync<T>(response);
        }

        private async Task<T> HandleResponseAsync<T>(HttpResponseMessage response)
        {
            var content = await response.Content.ReadAsStringAsync();
            if (!response.IsSuccessStatusCode)
            {
                throw new AntybrowserException($"API request failed with status {response.StatusCode}", response.StatusCode, content);
            }

            try
            {
                return JsonSerializer.Deserialize<T>(content, new JsonSerializerOptions
                {
                    PropertyNameCaseInsensitive = true
                }) ?? throw new AntybrowserException("Failed to deserialize response");
            }
            catch (JsonException ex)
            {
                throw new AntybrowserException("Invalid JSON response from API", ex);
            }
        }
        #endregion

        /// <summary>
        /// Disposes the underlying HTTP client.
        /// </summary>
        public void Dispose()
        {
            _httpClient.Dispose();
        }
    }
}

File: swift/Sources/AntybrowserClient.swift
import Foundation

public class AntybrowserClient {
    private let apiKey: String
    private let baseUrl: String
    private let session: URLSession
    private let decoder = JSONDecoder()
    
    public init(apiKey: String, port: Int = 5173) {
        self.apiKey = apiKey
        self.baseUrl = "http://127.0.0.1:\(port)"
        self.session = URLSession.shared
        decoder.keyDecodingStrategy = .convertFromSnakeCase
    }
    
    // System
    public func getStatus() async throws -> ApiResponse { try await get("/api/status") }
    public func getSettings() async throws -> Settings { try await get("/api/settings") }
    public func getSyncStatus() async throws -> SyncStatus { try await get("/api/sync/status") }
    
    public func refreshSync(profileId: String? = nil) async throws -> ApiResponse {
        let body: [String: Any] = profileId != nil ? ["profileId": profileId!] : [:]
        return try await post("/api/sync/refresh", body: body)
    }
    
    // Profiles
    public func getProfiles() async throws -> [Profile] { try await get("/api/profiles") }
    
    public func createProfile(_ request: CreateProfileRequest) async throws -> Profile {
        try await post("/api/profiles", body: request)
    }
    
    public func updateProfile(id: Int, data: [String: Any]) async throws -> Profile {
        try await put("/api/profiles/\(id)", body: data)
    }
    
    public func deleteProfile(id: Int) async throws -> ApiResponse {
        try await delete("/api/profiles/\(id)")
    }
    
    public func startProfile(id: Int) async throws -> StartProfileResponse {
        try await post("/api/profiles/\(id)/start", body: [:])
    }
    
    public func stopProfile(id: Int) async throws -> ApiResponse {
        try await post("/api/profiles/\(id)/stop", body: [:])
    }
    
    public func duplicateProfile(id: Int, name: String? = nil) async throws -> Profile {
        let body: [String: Any] = name != nil ? ["name": name!] : [:]
        return try await post("/api/profiles/\(id)/duplicate", body: body)
    }
    
    // Automations
    public func getAutomations() async throws -> [Automation] { try await get("/api/automations") }
    
    public func runAutomation(id: Int, profileId: Int) async throws -> RunAutomationResult {
        try await post("/api/automations/\(id)/run", body: ["profileId": profileId])
    }
    
    // Groups
    public func getGroups() async throws -> [Group] { try await get("/api/groups") }
    public func createGroup(_ request: CreateGroupRequest) async throws -> Group { try await post("/api/groups", body: request) }
    public func updateGroup(id: Int, data: [String: Any]) async throws -> Group { try await put("/api/groups/\(id)", body: data) }
    public func deleteGroup(id: Int) async throws -> ApiResponse { try await delete("/api/groups/\(id)") }
    
    // Proxies
    public func getProxies() async throws -> [Proxy] { try await get("/api/proxies") }
    public func createProxy(_ request: CreateProxyRequest) async throws -> Proxy { try await post("/api/proxies", body: request) }
    public func checkProxy(data: [String: Any]) async throws -> ProxyCheckResult { try await post("/api/proxies/check", body: data) }
    public func deleteProxy(id: Int) async throws -> ApiResponse { try await delete("/api/proxies/\(id)") }
    
    // Extensions
    public func getExtensions() async throws -> [Extension] { try await get("/api/extensions") }
    public func deleteExtension(id: Int) async throws -> ApiResponse { try await delete("/api/extensions/\(id)") }
    public func getProfileExtensions(profileId: Int) async throws -> [Extension] { try await get("/api/profiles/\(profileId)/extensions") }
    public func setProfileExtensions(profileId: Int, extensionIds: [Int]) async throws -> ApiResponse {
        try await post("/api/profiles/\(profileId)/extensions", body: ["extensionIds": extensionIds])
    }
    
    // HTTP helpers
    private func get<T: Decodable>(_ path: String) async throws -> T {
        let request = try makeRequest(path: path)
        let (data, response) = try await session.data(for: request)
        return try handleResponse(data: data, response: response)
    }
    
    private func post<T: Decodable>(_ path: String, body: Any) async throws -> T {
        var request = try makeRequest(path: path)
        request.httpMethod = "POST"
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await session.data(for: request)
        return try handleResponse(data: data, response: response)
    }
    
    private func put<T: Decodable>(_ path: String, body: Any) async throws -> T {
        var request = try makeRequest(path: path)
        request.httpMethod = "PUT"
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await session.data(for: request)
        return try handleResponse(data: data, response: response)
    }
    
    private func delete<T: Decodable>(_ path: String) async throws -> T {
        var request = try makeRequest(path: path)
        request.httpMethod = "DELETE"
        let (data, response) = try await session.data(for: request)
        return try handleResponse(data: data, response: response)
    }
    
    private func makeRequest(path: String) throws -> URLRequest {
        guard let url = URL(string: "\(baseUrl)\(path)") else { throw AntybrowserError.invalidURL }
        var request = URLRequest(url: url)
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return request
    }
    
    private func handleResponse<T: Decodable>(data: Data, response: URLResponse) throws -> T {
        guard let httpResponse = response as? HTTPURLResponse else { throw AntybrowserError.invalidResponse }
        guard (200...299).contains(httpResponse.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw AntybrowserError.apiError(statusCode: httpResponse.statusCode, body: body)
        }
        return try decoder.decode(T.self, from: data)
    }
}
